### PR TITLE
Implement mscomp using shelving filters #26

### PR DESCRIPTION
**Pull request by x42**
@magnetophon please review

### CODE REVIEW

**Comment by x42** [comment]
Sadly I started work before c490ac8a0fcb005b1f0ef94c38ef128bc3903a25. So parameter IDs need updating

**Comment by x42** [comment]
There is some oddity that I cannot [yet] explain
```
: par(i, N * Nr_bands, ba.linear2db : meter(i+1)); // XXX why does this produce only 15 meters, instead of 16 ?
```

N = 2; Nr_bands = 8;

**Comment by falkTX** [comment]
> There is some oddity that I cannot [yet] explain
> 
> ```
> : par(i, N * Nr_bands, ba.linear2db : meter(i+1)); // XXX why does this produce only 15 meters, instead of 16 ?
> ```
> 
> N = 2; Nr_bands = 8;

I had some oddities with that too, but then was not sure if it was in faust or me doing things wrong with the enums.
So currently the code has a +1

```
static_assert(kParameter_86 + 1 - kParameter_69 == 18, "has 18 multiband compressor passive controls");
```

But on this case, it is just a silliness of counting. if kParameter_86 itself was the only meter, this would make `kParameter_86 - kParameter_86 == 0` so... 0 meters?

**Comment by x42** [comment]
Looking at the .ttl, I get meter 10..16. followed by 1 ..8  . Number 9 is AWOL.
How can that be?

**Comment by magnetophon** [comment]
There was a bug in ``groupGainOut`` that led to the following connections:
image

This fixes it:

```
    groupGainOut  =
        (ro.interleave(Nr_bands,N), si.bus(N))
        :(ro.interleave(N, Nr_bands+1))
        :par(i, N, ro.cross(Nr_bands),_);
```

File: soundsgood.dsp
Comment (magnetophon) [inline soundsgood.dsp]: I already wrote that code:     
```faust
flist = LogArray(Nr_bands-1,fl,fh);
```

That way it also works for any number of bands.

File: soundsgood.dsp
Comment (magnetophon) [inline soundsgood.dsp]: You are going from db to linear and back.

I would put the following at the bottom of the file, use it here and get rid of the linear2dB before the meter:

```faust:
// +++++++++++++++ peak_compression_gain_mono_db ++++++++++++++++

// TODO: use co.peak_compression_gain_N_chan_db when it arrives in the current faust version

// see: https://github.com/grame-cncm/faustlibraries/pull/140, it's merged in faustlibraries master


peak_compression_gain_mono_db(strength,thresh,att,rel,knee,prePost) =
    abs : ba.bypass1(prePost,si.onePoleSwitching(att,rel)) : ba.linear2db : gain_computer(strength,thresh,knee) : ba.bypass1((prePost !=1),si.onePoleSwitching(rel,att))
with {
    gain_computer(strength,thresh,knee,level) =
        select3((level>(thresh-(knee/2)))+(level>(thresh+(knee/2))),
                0,
                ((level-thresh+(knee/2)) : pow(2)/(2*max(ma.EPSILON,knee))),
                (level-thresh))
        : max(0)*-strength;
```

When this change has trickled down to the current faust release, we can use: 
```faust
co.peak_compression_gain_N_chan_db
```
see: https://github.com/grame-cncm/faustlibraries/pull/140, it's merged in faustlibraries master

**Review by magnetophon — CHANGES_REQUESTED** [review]

File: soundsgood.dsp
Comment (x42) [reply soundsgood.dsp]: Thanks, done. 
I see you already have used the same in the zero-layency brickwall limiter. Perhaps after this is merged we can move it into common location, or even a separate file that's included on demand?

**Comment by x42** [comment]
@sletz, @magnetophon  how does one terminate  `%variablename` (without using a white-space char)?
say I have 2 variables  `foo(i,j)  = vbargraph("Foo [symbol:foo_%i_%j]")` --  In this case the first is interpreted as variable `i_`.

https://faustdoc.grame.fr/manual/syntax/#variable-parts-of-a-label also says that literal % produced by writing `%%`. That seems relevant for `"[unit:%]"` in various  labels.

File: soundsgood.dsp
Comment (magnetophon) [inline soundsgood.dsp]: I think you want:
```faust
: par(b, Nr_bands, par(c, N, ba.linear2db : meter(b+1, c+1)));
```
That way you get 8 times MS, like in the old version.

File: soundsgood.dsp
Comment (magnetophon) [inline soundsgood.dsp]: This then becomes:
```faust
  meter(c,b) = _<: attach(_, (max(-3):min(0):vbargraph("v:soundsgood/t:expert/h:[6]mscomp_meter/[%b.%c][unit:dB][tooltip: gain reduction in db][symbol:msredux%b%c]mscomp redux band %b chn %c", -3, 0)));
```

**Comment by sletz** [comment]
@x42 does this help? https://faustdoc.grame.fr/manual/syntax/#variable-parts-of-a-label

**Review by magnetophon — CHANGES_REQUESTED** [review]

**Comment by x42** [comment]
> @x42 does this help? https://faustdoc.grame.fr/manual/syntax/#variable-parts-of-a-label

No :)

File: soundsgood.dsp
Comment (x42) [reply soundsgood.dsp]: That'd be incorrect order. the gain parameters are order 
  8 values for  leftchannel,  8 values for right channel
hence  `par(c, N, par(b, Nr_bands, meter(c+1, b+1))); `

File: soundsgood.dsp
Comment (x42) [reply soundsgood.dsp]: yes, the meter definition can change, yes. the calls to meter(c,b) cannot.

Comment (x42) [reply soundsgood.dsp]: Done. Now the order is the same as before.
It'd be kinda cool to also include the actual freq-band range in the label or tooltip.

**Comment by x42** [comment]
> > @x42 does this help? https://faustdoc.grame.fr/manual/syntax/#variable-parts-of-a-label
> 
> No :)

@sletz  The manual does not document what characters are valid for variable names, nor how to separate variable names.  -- I tried bash like syntax `%{i}` and GNU style `%(i)` to no avail.

File: soundsgood.dsp
Comment (magnetophon) [reply soundsgood.dsp]: > yes, the meter definition can change, yes. the calls to meter(c,b) cannot.

[This commit](https://github.com/trummerschlunk/soundsgood/commit/e93c2e74ae2ccc2336d8fe22966e955ddb39d2c8) gives this routing:

image

Each compressor block is an N channel compressor, where N =2 in this case.
The labels do not mach the actual channels.  

When you turn up both link parameters, you can also see that the wrong meters are linked. 

> It'd be kinda cool to also include the actual freq-band range in the label or tooltip.

That woud be great, but I don't think faust can do dynamic labels, right @sletz?
So it would become a job for the C++ programmers in the team. :)

**Comment by sletz** [comment]
> first is interpreted as variable `i_`

Seems like a bug.

File: soundsgood.dsp
Comment (magnetophon) [reply soundsgood.dsp]: @x42 What does [your latest commit](https://github.com/trummerschlunk/soundsgood/compare/e93c2e74ae2ccc2336d8fe22966e955ddb39d2c8..67e0b526f616f4e1a3d731215c8f2f77f7d996e9) do?
So far I've been only testing the faust code, as I haven't gotten around to making a nix build for the GUI yet.

Comment (magnetophon) [reply soundsgood.dsp]: NM, it's putting back the bypass in the brickwall, isn't it?
It looked like it was changing the meters, but that's just because a parameter got added, right?

Comment (x42) [reply soundsgood.dsp]: Yes rebased to master - I'm testing with `faust2lv2` and the parameters are matched to master branch.
What is the issue with compressor meter ordering?

**Comment by x42** [comment]
> > first is interpreted as variable `i_`
> 
> Seems like a bug.

should I report this, or can you take it from here?

File: soundsgood.dsp
Comment (magnetophon) [inline soundsgood.dsp]: @x42 Turns out we where both kinda right: the order of the meters was correct in relation to the bands in your version.
However, that is not the order that the compressors need the bands in, in oder to do proper linking of the channels.

To have everything correct, you need:
```faust
  comps(N) = (strength_array, thresh_array, att_array, rel_array, knee_array, link_array, si.bus(N*Nr_bands))
             : ro.interleave(Nr_bands,6+N)
             : par(i, Nr_bands, compressor(N,prePost)) // : si.bus (N * Nr_bands)
             : par(c, N, par(b, Nr_bands, meter(c+1, b+1))):ro.interleave (N,Nr_bands);
```

**Review by magnetophon — CHANGES_REQUESTED** [review]

**Comment by sletz** [comment]
> should I report this, or can you take it from here?

See the definition of what is considered part of the identifier:
https://github.com/grame-cncm/faust/blob/master-dev/compiler/evaluate/eval.cpp#L794

Not sure if its a bug or a "non documented" feature ((-; yes please report it so that @orlarey can react

File: soundsgood.dsp
Comment (x42) [reply soundsgood.dsp]: Thanks for noticing that. pushed
For some reason I find  `ro.interleave` most confusing.

Comment (magnetophon) [reply soundsgood.dsp]: @x42 I made a mistake in my original comment: I didn't interleave back!
My edit crossed your push, timing-wise; sorry! 

> For some reason I find ro.interleave most confusing.

You know what it does, right?
If not, you should look at the block diagrams more!
Very handy for checking your work; it's how I found most of the issues in this PR.

**Comment by x42** [comment]
> Not sure if its a bug or a "non documented" feature ((-; yes please report it so that @orlarey can react

https://github.com/grame-cncm/faust/issues/787

File: soundsgood.dsp
Comment (x42) [reply soundsgood.dsp]: Ah, yeah, I forgot to move the interleave to the end. Thanks.

Perhaps it is easier to use a branch on the main repo and you can continue there (rather than using a fork, since I had no write access)?

Comment (magnetophon) [reply soundsgood.dsp]: > Perhaps it is easier to use a branch on the main repo and you can continue there (rather than using a fork, since I had no write access)?

Yeah, how do we do that?
Shall I start a new PR?

Comment (magnetophon) [reply soundsgood.dsp]: Cause I just noticed I told you the wrong thing again; there's still a bug, sorry!
Working on it...

Comment (x42) [reply soundsgood.dsp]: The *original* code (after fixing the `groupGainOut`) worked correctly. I checked by sending sine waves of various freq through it. Now there's indeed something vexxed.

Comment (magnetophon) [reply soundsgood.dsp]: > The _original_ code (after fixing the `groupGainOut`) worked correctly. I checked by sending sine waves of various freq through it. Now there's indeed something vexxed.

Except the channel linking was borked. :)

I think I worked it out now, just testing a but more.
Shall I do a new PR when I'm done?

Comment (x42) [reply soundsgood.dsp]: BTW, do you find the interpolated "strength, threshold, attack, release and link" parameters useful?
I guess that's also a question for @trummerschlunk - have you done listening tests tweaking those params laid out as-is?

Comment (x42) [reply soundsgood.dsp]: > I think I worked it out now, just testing a but more.

Great, thanks for looking into this!

> Shall I do a new PR when I'm done?

Sure, whatever is easiest for you.

Comment (magnetophon) [reply soundsgood.dsp]: > BTW, do you find the interpolated "strength, threshold, attack, release and link" parameters useful? I guess that's also a question for @trummerschlunk - have you done listening tests tweaking those params laid out as-is?

I think it's way better than the alternatives:
- each parameters 8 times
- each parameter once, the same for each compressor
- each parameter once, with hardcoded offsets
- all parameters hardcoded

The first option is great when you want to manually fix certain problems in a mix, but when you are making presets that need to sound good on anything, you want to have the parameters interpolated like this anyway.

We might want to try and have offsets that cater to Fletcher–Munson for the threshold, strength and/or knee.
I am doing [something similar](https://github.com/magnetophon/VoiceOfFaust/blob/82c20ed40fb5f184a236a7c765eab5acfc3bb5f5/lib/general.lib#L249-L277) in Voice of Faust.
With one knob, you can crossfade the bands between no weighting, A-weighting and ITU-R 468 weighting.

@x42 How would you do it?

Comment (x42) [reply soundsgood.dsp]: I'd use 
* A common release time for all (no interpolation) bands.
* A single attack value that is however scaled for each freq band (by 1/freq, to be at least 1 period long).
* strengths per band.  -- I think you may not want to to compress bass much (very audible in music),  work on the low and high mids separately for voices (de-ess), and the top also depends. cymbals or ..?
* likewise (relative) thresholds per band.
* knee can be identical for all.
* I don't know how "link" works in a m/s context. It is kind weird, there also comb filter like artifacts when the side limiter kicks in  (generally speaking L/R should be linked, but that doesn't apply here). -- I played with the current controls without being anythewiser while listening.

**Comment by orlarey** [comment]
In label strings, "....%i...." identifier names, like 'i', have to be one
character long. By writing "....%i_...." you create a
two-character identifier because _ is not a separator but is among the
allowed identifier characters. Instead, you could write  "....%i,....".

Le lun. 22 août 2022 à 20:35, Robin Gareus ***@***.***> a
écrit :

File: soundsgood.dsp
Comment (x42) [reply soundsgood.dsp]: I think the main issue is that there is linear interpolation low -> high.

For dialog, I compress the bass and highs a lot, while leaving the mids alone. For music usually the opposite.
So a "U" shape would work better, and a control (like gamma for brightness) that can turn it into an inverted U?
Also 2 controls only, just not linear interpolation?

**Comment by x42** [comment]
> "....%i,....".

sadly that won't work here. The generated **symbol** name  is used in C/C++ GUI code and needs to be a valid C variable name. commas are not allowed there.  here e.g. "meter_2_3"

File: soundsgood.dsp
Comment (x42) [reply soundsgood.dsp]: anyway. let's first focus on getting the mscomp with shelves done and worry about this later. It's not release critical

Comment (magnetophon) [reply soundsgood.dsp]: > * A common release time for all (no interpolation) bands.
> * A single attack value that is however scaled for each freq band (by 1/freq, to be at least 1 period long).

You wouldn't want the release time to be relate to the period?
I find I get way cleaner compression with longer release in the lows and shorter in the highs.

The logarithmic interpolation I used here allows for both, with an easy to understand UI.   

> * strengths per band. -- I think you may not want to to compress bass much (very audible in music), work on the low and high mids separately for voices (de-ess), and the top also depends. cymbals or ..?
> * likewise (relative) thresholds per band.

That's a lot of parameters!

>For dialog, I compress the bass and highs a lot, while leaving the mids alone. For music usually the opposite.
So a "U" shape would work better, and a control (like gamma for brightness) that can turn it into an inverted U?
Also 2 controls only, just not linear interpolation?

So for speech an U shape and for music an inverted U?
So for general purpose a line?  ;)

More seriously: For Voice of Faust I also made an interpolator that has a parametric mid: you set the value you want the middle band to have, and you set the number of the band that is the middle.  

> I don't know how "link" works in a m/s context. 

When you unlink you usually mostly compress the middle, cause it's louder, so you get a wider stereo image.
When you link, you keep the stereo image as it is. 

> It is kind weird, there also comb filter like artifacts when the side limiter kicks in 

I haven't heard those. Can you link a track that has this issue with the default settings? 

> I played with the current controls without being anythewiser while listening.

Have you played with the version from this PR or from master?

**Comment by falkTX** [comment]
> > "....%i,....".
> 
> sadly that won't work here. The generated **symbol** name is used in C/C++ GUI code and needs to be a valid C variable name. commas are not allowed there. here e.g. "meter_2_3"

One ugly but possible solution is using $ as character separator.
It is allowed in both GCC and MSVC, see https://stackoverflow.com/questions/7926394/dollar-sign-in-variable-name
And per C++11 onwards we are allowed to use more than just a-zA-Z0-9_ chars.

**Comment by x42** [comment]
> One ugly but possible solution is using $ as character separator.

nice trick. For now we're good  "%b%c" works since it satisfies the prefix property with n_bands = 8, n_channels=2.
It may however become relevant for other projects.

File: soundsgood.dsp
Comment (x42) [reply soundsgood.dsp]: I've mainly used a "made with Ardour" track https://youtu.be/vzdebCx4Kzg
at around 1:50 there are hand-claps + vocals.  Not quite what you get with an applauding audience, but similarly mean. As for m/s this is noticeable on the piano and strings (and reverb).

I did some A/B earlier. The phase issues (fixed by the shelves) is less dramatic than I anticipated.
Even though the song is already mastered it, the loudness increases over time and the range becomes smaller (climax at the end). Also at the end hand-claps are L/R panned and there's a 2nd snare on the right channel.

Comment (x42) [reply soundsgood.dsp]: any progress? From what I can tell the earlier version
```
  comps(N) = (strength_array, thresh_array, att_array, rel_array, knee_array, link_array, si.bus(N*Nr_bands))
    : ro.interleave(Nr_bands,6+N)
    : par(i, Nr_bands, compressor(N,prePost)) // : si.bus (N * Nr_bands)
    : par(b, Nr_bands, par(c, N, meter(c+1, b+1)));
 ```
is correct.

Comment (magnetophon) [reply soundsgood.dsp]: @x42 Are you still hearing issues with https://github.com/trummerschlunk/soundsgood/pull/38?
I listen to your track on the default settings, and it sounded fine.

**Comment by x42** [comment]
OK, I'll close this in favor of #38

Thanks for all the input and feedback! Onward

**Pull request closed without merging**
